File: recursion/myAtoi.cpp
#include<bits/stdc++.h>
using namespace std;

int atoiHelper(string s, int idx, int result) {
    // base case
    if (idx == s.size()) return result;

    char c = s[idx];
    if (!isdigit(s[idx])) return result;
    int digit = s[idx] - '0';
    return atoiHelper(s, idx+1, result*10+digit);
}

int myAtoi(string s) {    
    int result = 0;
    int i = 0, sign = 1;

    // skip whitespaces
    while (i < s.size() && s[i] == ' ' ) i++;

    // handle sign
    if ( i < s.size() && (s[i] == '+' || s[i] == '-')) {
        sign = s[i] == '-' ? -1 : 1;
        i++;
    }

    result = atoiHelper(s,i,result);

    return result * sign;
}

int main()
{
    string s = "1234";
    cout << myAtoi(s);
    return 0;
}

File: ruby/recursion/my_atoi.rb
# def my_atoi(str)
#   an = 0
#   str = str.strip
#   flag = str[0] == '-' ? false : true
    
#   str.each_char do |char|
#     if char == ' '
#       break
#     else
#       converted_char =  char.to_i
#       an = an * 10 + converted_char
#     end
#   end
  
#   flag == false ? an = an * -1 : ''
#   an
# end

def atoi_helper(str, idx, res)
  # base condition
   if idx == str.length || str[idx] == ' ' || !str[idx].match?(/\d/)
    return res
   end

   num = str[idx].to_i 
   return atoi_helper(str, idx+1, res*10 + num)
end

def my_atoi(str)
  return nil if str.empty? 

  # trim
  str = str.strip 

  # handle sign
  signed_value = str[0] == '-' ? -1 : 1
  
  new_str = signed_value == -1 ? str[1..] : str
  result = atoi_helper(new_str, 0, 0)

  return result*signed_value

end

# num = "       -423  254      "
num = "432"
# num = "1337c0d3"
puts my_atoi(num)